Refactor Dockerfile.ci to slim containers

TrevorBenson opened the pull request:
## Content
Refactors the Dockerfile and Dockerfile.ci to use release binaries. 

## Pre-submit checklist

- Branch
  - [ ] Tests are provided (if possible)
  - [ ] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [X] Key commits have useful messages
- PR
  - [ ] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [X] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [ ] Update README file (if relevant)
  - [ ] Update documentation website (if relevant)
  - [ ] Add dev blog post (if relevant)

## Comments
* Reduces the size of existing containers. 
* Does not implement a BuildX cache to leverage identical image layers from previous CI runs.

## Issue(s)
Relates to #1272

TrevorBenson commented:
@jpraynaud @abailly-iohk 
I don't have access to add reviewers to the PR, thanks.

abailly-iohk commented:
@TrevorBenson No worries, thanks a lot for the PR! We'll review it ASAP

TrevorBenson commented:
Adding this to the PR discussion so it doesn't get overlooked. Potential to merge Dockerfiles into a single file for both use cases: https://github.com/input-output-hk/mithril/issues/1272#issuecomment-1764993870

jpraynaud commented on mithril-aggregator/Dockerfile.ci:
```diff
@@ -3,25 +3,32 @@
 RUN /app/bin/cardano-cli --version
 RUN rm -f cardano-bin.tar.gz
 
+# Install mithril-aggregator
+ARG MITHRIL_BIN_URL=https://github.com/input-output-hk/mithril/releases/download/2337.0/mithril-2337.0-linux-x64.tar.gz
+RUN wget -nv -O mithril-bin.tar.gz $MITHRIL_BIN_URL \
+    && tar xzf mithril-bin.tar.gz mithril-aggregator && mv mithril-aggregator /app/bin && chmod +x /app/bin/mithril-aggregator \
+    && /app/bin/mithril-aggregator --version \
+    && rm -f mithril-bin.tar.gz
```
This Dockerfile is used by the CI to package a pre-compiled binary, so this should be rolled-back to the previous command:
```suggestion
# Copy the executable
COPY mithril-aggregator/mithril-aggregator /app/bin/mithril-aggregator
```

jpraynaud commented on mithril-client/Dockerfile.ci:
```diff
@@ -3,15 +3,24 @@
 
-# Create appuser
-RUN adduser --disabled-password appuser
+# Install mithril-client
+ARG MITHRIL_BIN_URL=https://github.com/input-output-hk/mithril/releases/download/2337.0/mithril-2337.0-linux-x64.tar.gz
+RUN wget -nv -O mithril-2337.0-linux-x64.tar.gz $MITHRIL_BIN_URL \
+    && tar xzf mithril-2337.0-linux-x64.tar.gz mithril-client && mv mithril-client /app/bin && chmod +x /app/bin/mithril-client \
+    && /app/bin/mithril-client --version \
+    && rm -f mithril-2337.0-linux-x64.tar.gz
```
Same comments as above for this image:
```suggestion
# Copy the executable
COPY mithril-client/mithril-client /app/bin/mithril-client
```

jpraynaud commented on mithril-signer/Dockerfile.ci:
```diff
@@ -1,26 +1,33 @@
+    && rm -f cardano-bin.tar.gz
+
+# Install mithril-signer
+ARG MITHRIL_BIN_URL=https://github.com/input-output-hk/mithril/releases/download/2337.0/mithril-2337.0-linux-x64.tar.gz
+RUN wget -nv -O mithril-bin.tar.gz $MITHRIL_BIN_URL \
+    && tar xzf mithril-bin.tar.gz mithril-signer && mv mithril-signer /app/bin && chmod +x /app/bin/mithril-signer \
+    && /app/bin/mithril-signer --version \
+    && rm -f mithril-bin.tar.gz
```
Same also here:
```suggestion
# Copy the executable
COPY mithril-signer/mithril-signer /app/bin/mithril-signer
```

jpraynaud commented on mithril-aggregator/Dockerfile:
As you mentioned, this file and all the other `Dockerfile` should not be modified :slightly_smiling_face:

TrevorBenson commented:
1. Dropped the commit with changes to each `Dockerfile`.
2. Refactored `Dockerfile.ci` files to:
   * Slim the **Install cardano-cli** step.
   * Keep the COPY for installation of any mithril binary.
   * Optimize the order of steps to reduce cache invalidation.

### Cache Invalidation/Optimizations
* Developers will see cache optimizations when they build images for their local registry from Dockerfile.ci.
* The ci/release workflows build and push Action won't take advantage of the cache optimizations without the buildx cache implemented in the workflow.
  * When the **debian:11-slim** base layer does not change developers pull every layer after the base from ghcr.io
  * When the **debian:11-slim** base layer changes developers will pull every layer from ghcr.io

jpraynaud commented on mithril-aggregator/Dockerfile.ci:
```diff
@@ -3,24 +3,27 @@
-RUN /app/bin/cardano-cli --version
-RUN rm -f cardano-bin.tar.gz
+# Copy the config files
+COPY config /app/config
```
It looks like some [tests failed in the CI](https://github.com/input-output-hk/mithril/actions/runs/6550181649/job/17789702156#step:6:186) because the config directory is incorrect: 
```suggestion
COPY mithril-aggregator/config /app/config
```

jpraynaud commented on mithril-client/Dockerfile.ci:
```diff
@@ -3,15 +3,20 @@
 # Copy the executable
 COPY mithril-client/mithril-client /app/bin/mithril-client
-COPY mithril-client/config /app/config
+
+# Copy the config files
+COPY config /app/config
```
Same problem as above:
```suggestion

# Copy the config files
COPY mithril-client/config /app/config
```

TrevorBenson replied:
Yep, I saw the failure and already adjusted for it. New checks should kick off after workflow approval

jpraynaud commented on mithril-signer/Dockerfile.ci:
```diff
@@ -1,26 +1,29 @@
-RUN tar xzf cardano-bin.tar.gz ./cardano-cli && mv cardano-cli /app/bin
-RUN /app/bin/cardano-cli --version
-RUN rm -f cardano-bin.tar.gz
+# Copy the config files
+COPY config /app/config
```
Same also here:
```suggestion
# Copy the config files
COPY mithril-signer/config /app/config
```

jpraynaud reviewed (commented):
Some [CI tests](https://github.com/input-output-hk/mithril/actions/runs/6550181649) have failed, here are some changes that should make them green :+1:

TrevorBenson commented:
@jpraynaud The only two failures now are for **ci/hydra-build:aarch64-darwin.mithril-signer** and **ci/hydra-build:x86_64-darwin.mithril-signer**, both of which I observed fail in other PR branches, so suspect its unrelated.

Alenar reviewed (approved):
Looks good to me, thanks :rocket: !

jpraynaud reviewed (approved):
LGTM :+1:

jpraynaud commented:
@TrevorBenson you need to sign your commits in order to get them merged in the repository.
Can you re-push them?

jpraynaud commented:
> > @TrevorBenson you need to sign your commits in order to get them merged in the repository.
> > Can you re-push them?
> 
> Probably worth adding this to the contribution guidelines. I don't recall seeing it when I initially reviewed.

You're right, I have created this PR https://github.com/input-output-hk/mithril/pull/1305 to add a mention to this requirement :+1:

TrevorBenson commented:
> @TrevorBenson you need to sign your commits in order to get them merged in the repository. Can you re-push them?

Commit signed.

TrevorBenson commented:
I was looking into the new, third, failed check so rebased onto the main branch and pushed again. I didn't catch that you had already re-run the failed job and got a success. :shrug: :smile:

jpraynaud commented on mithril-signer/Dockerfile.ci:
```diff
@@ -1,26 +1,29 @@
 # Creates a docker image to run an executable built outside of the image
-# This relies on the fact the mithril-signer executable has been built
+# This relies on the fact the mithril-client executable has been built
```
@TrevorBenson could you fix this small typo?
```suggestion
# This relies on the fact the mithril-signer executable has been built
```

TrevorBenson replied:
Would you like a new commit with "review suggestions" in the description, or a fixup?

jpraynaud replied:
A fixup is fine :+1: 
Thanks!

TrevorBenson replied:
Done

jpraynaud commented:
Many thanks @TrevorBenson for your contribution!
